### PR TITLE
feat: forbid modify a shared table

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

1. add `share_name` into `TableInfo`.
2. forbid modify a shared table if `TableInfo.share_name` is_some.
3. add more `check_mutable` in write table operations.
4. remove unused `TableInfo` convert to pb code.

Fixes https://github.com/datafuselabs/databend/issues/7641

### CODE REVIEW

File: src/meta/proto-conv/src/table_from_to_protobuf_impl.rs [hunk 1]
```diff
@@ -78,43 +78,6 @@ impl FromToProto for mt::TableCopiedFileLock {
     }
 }
 
-impl FromToProto for mt::TableInfo {
```
Comment (drmingdrmer) [inline src/meta/proto-conv/src/table_from_to_protobuf_impl.rs:81]: Nice! Love to see codes being removed

**Review by drmingdrmer — APPROVED** [review]
It looks great to me!

**Comment by BohuTANG** [comment]
@mergify update

**Comment by Xuanwo** [comment]
> CI looks like something is wrong. cc @Xuanwo

Yep. It will be fixed in https://github.com/datafuselabs/databend/pull/7681

**Comment by Xuanwo** [comment]
@mergify update

**Comment by BohuTANG** [comment]
```
Following failed statements:
---------------------------------------------
Runner: mysql
ErrorType: statement ok execute with exception
Message: Failed to execute. Collected info: 1105 (HY000): Code: 1025, displayText = Unknown table '05_0003_at_t0'.

Parsed Statement
    at_line: 25,
    s_type: Statement: ok, type: None,
    suite_name: base/05_ddl/05_0003_ddl_alter_table,
    text:
        ALTER TABLE IF EXISTS `05_0003_at_t0` RENAME TO `05_0003_at_t1`;
    results: [],
    runs_on: {'http', 'clickhouse', 'mysql'},
---------------------------------------------
Runner: http
ErrorType: statement ok get error in response
Message: 
Parsed Statement
    at_line: 25,
    s_type: Statement: ok, type: None,
    suite_name: base/05_ddl/05_0003_ddl_alter_table,
    text:
        ALTER TABLE IF EXISTS `05_0003_at_t0` RENAME TO `05_0003_at_t1`;
    results: [],
    runs_on: {'http', 'clickhouse', 'mysql'},
```

File: src/query/service/src/interpreters/interpreter_cluster_key_alter.rs [hunk 2]
```diff
@@ -62,6 +62,8 @@ impl Interpreter for AlterTableClusterKeyInterpreter {
             .get_table(tenant.as_str(), &plan.database, &plan.table)
             .await?;
```
Comment (BohuTANG) [inline src/query/service/src/interpreters/interpreter_cluster_key_alter.rs:64]: Why we check_modify for cluster operation?

Comment (lichuang) [reply src/query/service/src/interpreters/interpreter_cluster_key_alter.rs:64]: I am not quite sure, cause it is a type operation of `AlterTableStmt`, so i add `check_can_modify_table` in every `AlterTableStmt` operation, it forbid modify the table if it is a shared table.

@dantengsky have a look?

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_cluster_key_alter.rs:64]: I think we should add a new share access check like management mode check here: https://github.com/datafuselabs/databend/blob/f2a8f0da2e966376b954bf7e357ff0c4ce86704f/src/query/service/src/interpreters/access/management_mode_access.rs

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_cluster_key_alter.rs:64]: It seems we should refactoring the access in InterpreterFactory to a common access instead the ManagementAccess.
Access.check():
management_access.check()?(
shared_table_access.check()?

Comment (BohuTANG) [reply src/query/service/src/interpreters/interpreter_cluster_key_alter.rs:64]: Open a PR for it in https://github.com/datafuselabs/databend/pull/7687
Then we can register a new checker in Accessor for the shared_table_accessor.

File: src/query/catalog/src/table.rs [hunk 3]
```diff
@@ -71,6 +71,15 @@ pub trait Table: Sync + Send {
 
     fn get_table_info(&self) -> &TableInfo;
 
+    fn check_can_modify_table(&self) -> Result<()> {
```
Comment (BohuTANG) [inline src/query/catalog/src/table.rs:74]: We can move this check to a privilege check?
https://github.com/datafuselabs/databend/pull/7721

The prepared work is you should bind the share_name to table info in the binder.

**Comment by Xuanwo** [comment]
Replaced by https://github.com/datafuselabs/databend/issues/7818

**Pull request closed without merging**